modules/starport/src/lib.rs

```diff
@@ -0,0 +1,367 @@
+		//#[pallet::weight(< T as Config >::WeightInfo::lock_to())]
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn lock_to(
+			origin: OriginFor<T>,
+			to: T::AccountId,
+			currency_id: CurrencyId,
+			locked_amount: Balance,
+		) -> DispatchResultWithPostInfo {
+			let from = ensure_signed(origin)?;
+			Self::do_lock_to(from, to, currency_id, locked_amount)
+		}
+
+		/// Invoke a Notice issued from Compound chain via its Gateways.
+		///
+		/// Parameters:
+		/// - `notice`: The Notice issued by Compound Gateway. Contains data to be invoked.
+		/// - `signatures`: Represents approvals by given authorities. Used to verify the
+		/// authenticity of the notice.
+		//#[pallet::weight(< T as Config >::WeightInfo::invoke())]
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn invoke(
+			origin: OriginFor<T>,
+			notice: GatewayNotice<T::AccountId>,
+			signatures: Vec<CompoundAuthoritySignature>,
+		) -> DispatchResultWithPostInfo {
+			let from = ensure_signed(origin)?;
+
+			// Invoke can only be called from specific Gateway admin account
+			ensure!(T::GatewayAccount::get() == from, Error::<T>::InvalidNoticeInvoker);
+
+			// Calculate the hash for this notice, and ensure it is only invoked once.
+			let hash = BlakeTwo256::hash(&(serde_json::to_string(&notice).unwrap().as_bytes())[..]);
+
+			ensure!(
+				!InvokedNoticeHashes::<T>::contains_key(&hash),
+				Error::<T>::NoticeAlreadyInvoked
+			);
+
+			// verify the signatures
+			ensure!(
+				Self::verify_compound_authority_signature(signatures),
+				Error::<T>::InsufficientValidNoticeSignatures
+			);
+
+			match notice {
+				GatewayNotice::SetSupplyCap(currency_id, amount) => {
+					SupplyCaps::<T>::insert(&currency_id, amount.clone());
+					Self::deposit_event(Event::<T>::SupplyCapSet(currency_id, amount));
+					Ok(().into())
+				}
+				GatewayNotice::ChangeAuthorities(new_authorities) => {
+					ensure!(
+						new_authorities.len() <= (T::MaxGatewayAuthorities::get() as usize),
+						Error::<T>::ExceededMaxNumberOfAuthorities
+					);
+					let bounded_vec = BoundedVec::try_from(new_authorities).unwrap();
+					GatewayAuthorities::<T>::put(bounded_vec);
+					Self::deposit_event(Event::<T>::GatewayAuthoritiesChanged);
+					Ok(().into())
+				}
+				GatewayNotice::Unlock(currency_id, amount, who) => {
+					Self::do_unlock(currency_id.clone(), amount.clone(), who.clone())
+				}
+				GatewayNotice::SetFutureYield(next_cash_yield, yield_index, timestamp_effective) => {
+					T::Cash::set_future_yield(
+						next_cash_yield.clone(),
+						yield_index.clone(),
+						timestamp_effective.clone(),
+					)?;
+					Self::deposit_event(Event::<T>::FutureYieldSet(
+						next_cash_yield,
+						yield_index,
+						timestamp_effective,
+					));
+					Ok(().into())
+				}
+			}?;
+
+			// After its invocation, store the hash.
+			InvokedNoticeHashes::<T>::insert(&hash, ());
+
+			Ok(().into())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	#[transactional]
```

Comment:
no need to put `#[transactional]` in private inner help method. could use `#[require_transactional]`